File: Wave/src/protocol/codec.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * Tokenized-array codec — TS side.
 *
 * Encodes/decodes objects to compact positional arrays using the SHARED
 * `protocol/wire_manifest.json`. The Python codec (`server/wave_api/codec.py`)
 * is a line-for-line mirror, so the app and server decode identical arrays over
 * MQTT today and LoRa later. Field ORDER comes only from the manifest — never
 * hand-order anything here.
 */
import manifest from '@protocol/wire_manifest.json';

export type Token = string | number | boolean | null | Token[];

interface FieldDef {
  name: string;
  t: string;
  optional?: boolean;
}
interface TypeDef {
  tag?: number;
  fields: FieldDef[];
}

const MESSAGES = manifest.messages as Record<string, TypeDef>;
const DEFS = manifest.defs as Record<string, TypeDef>;
const ENUMS = manifest.enums as Record<string, string[]>;

export const PROTOCOL_VERSION = manifest.protocolVersion as number;

function resolveDef(name: string): TypeDef {
  const def = MESSAGES[name] ?? DEFS[name];
  if (!def) throw new Error(`codec: unknown type "${name}"`);
  return def;
}

const SCALARS = new Set(['str', 'int', 'bool']);

/** Encode a single value given its manifest type token. */
function encodeValue(t: string, value: any): Token {
  if (value === undefined || value === null) return null;

  if (SCALARS.has(t)) return value as Token;

  if (t === 'array') return value as Token; // raw passthrough (envelope payload)

  if (t.startsWith('enum:')) {
    const dict = ENUMS[t.slice(5)];
    const idx = dict.indexOf(value);
    if (idx < 0) throw new Error(`codec: value "${value}" not in enum ${t}`);
    return idx;
  }

  if (t.startsWith('array:')) {
    const elem = t.slice(6);
    return (value as any[]).map((v) => encodeValue(elem, v));
  }

  if (t.startsWith('map:')) {
    const elem = t.slice(4);
    return Object.entries(value as Record<string, any>).map(
      ([k, v]) => [k, encodeValue(elem, v)] as Token,
    );
  }

  // ref to a message/def
  return encodeFields(resolveDef(t).fields, value);
}

function decodeValue(t: string, token: Token): any {
  if (token === null || token === undefined) return undefined;

  if (SCALARS.has(t)) return token;

  if (t === 'array') return token; // raw passthrough

  if (t.startsWith('enum:')) {
    const dict = ENUMS[t.slice(5)];
    return dict[token as number];
  }

  if (t.startsWith('array:')) {
    const elem = t.slice(6);
    return (token as Token[]).map((v) => decodeValue(elem, v));
  }

  if (t.startsWith('map:')) {
    const elem = t.slice(4);
    const out: Record<string, any> = {};
    for (const pair of token as Token[]) {
      const [k, v] = pair as Token[];
      out[k as string] = decodeValue(elem, v);
    }
    return out;
  }

  return decodeFields(resolveDef(t).fields, token as Token[]);
}

export function encodeFields(fields: FieldDef[], obj: Record<string, any>): Token[] {
  return fields.map((f) => encodeValue(f.t, obj?.[f.name]));
}

export function decodeFields(fields: FieldDef[], arr: Token[]): Record<string, any> {
  const out: Record<string, any> = {};
  fields.forEach((f, i) => {
    const decoded = decodeValue(f.t, arr[i]);
    if (decoded !== undefined) out[f.name] = decoded;
  });
  return out;
}

/** Encode a named message/def into its positional token array. */
export function encode(typeName: string, obj: Record<string, any>): Token[] {
  return encodeFields(resolveDef(typeName).fields, obj);
}

/** Decode a positional token array back into an object for a named type. */
export function decode<T = Record<string, any>>(typeName: string, arr: Token[]): T {
  return decodeFields(resolveDef(typeName).fields, arr) as T;
}

export function typeTag(typeName: string): number {
  const tag = MESSAGES[typeName]?.tag;
  if (tag === undefined) throw new Error(`codec: "${typeName}" has no tag (not a top-level message)`);
  return tag;
}

export const ENVELOPE_FIELDS = (manifest.envelope as TypeDef).fields;

export function encodeEnvelope(env: Record<string, any>): Token[] {
  return encodeFields(ENVELOPE_FIELDS, env);
}

export function decodeEnvelope(arr: Token[]): Record<string, any> {
  return decodeFields(ENVELOPE_FIELDS, arr);
}

File: server/wave_api/codec.py
"""
Tokenized-array codec - Python side.

Line-for-line mirror of the TS codec (Wave/src/protocol/codec.py's sibling at
Wave/src/protocol/codec.ts). Both read the SAME protocol/wire_manifest.json so
the app and server encode/decode identical positional arrays over MQTT (and LoRa
later). Field ORDER comes only from the manifest.
"""
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

# Repo root: server/wave_api/codec.py -> parents[2] == Wave-1/
_MANIFEST_PATH = Path(__file__).resolve().parents[2] / "protocol" / "wire_manifest.json"

with _MANIFEST_PATH.open(encoding="utf-8") as fh:
    _MANIFEST = json.load(fh)

MESSAGES: dict[str, dict] = _MANIFEST["messages"]
DEFS: dict[str, dict] = _MANIFEST["defs"]
ENUMS: dict[str, list[str]] = _MANIFEST["enums"]
ENVELOPE_FIELDS: list[dict] = _MANIFEST["envelope"]["fields"]
PROTOCOL_VERSION: int = _MANIFEST["protocolVersion"]

_SCALARS = {"str", "int", "bool"}

Token = Any  # str | int | bool | None | list


def _resolve_def(name: str) -> dict:
    d = MESSAGES.get(name) or DEFS.get(name)
    if d is None:
        raise ValueError(f'codec: unknown type "{name}"')
    return d


def _encode_value(t: str, value: Any) -> Token:
    if value is None:
        return None

    if t in _SCALARS:
        return value

    if t == "array":  # raw passthrough (envelope payload)
        return value

    if t.startswith("enum:"):
        dict_ = ENUMS[t[5:]]
        try:
            return dict_.index(value)
        except ValueError:
            raise ValueError(f'codec: value "{value}" not in enum {t}')

    if t.startswith("array:"):
        elem = t[6:]
        return [_encode_value(elem, v) for v in value]

    if t.startswith("map:"):
        elem = t[4:]
        return [[k, _encode_value(elem, v)] for k, v in value.items()]

    # ref to a message/def
    return _encode_fields(_resolve_def(t)["fields"], value)


def _decode_value(t: str, token: Token) -> Any:
    if token is None:
        return None

    if t in _SCALARS:
        return token

    if t == "array":
        return token

    if t.startswith("enum:"):
        return ENUMS[t[5:]][token]

    if t.startswith("array:"):
        elem = t[6:]
        return [_decode_value(elem, v) for v in token]

    if t.startswith("map:"):
        elem = t[4:]
        return {k: _decode_value(elem, v) for k, v in token}

    return _decode_fields(_resolve_def(t)["fields"], token)


def _encode_fields(fields: list[dict], obj: dict) -> list[Token]:
    obj = obj or {}
    return [_encode_value(f["t"], obj.get(f["name"])) for f in fields]


def _decode_fields(fields: list[dict], arr: list[Token]) -> dict:
    out: dict[str, Any] = {}
    for i, f in enumerate(fields):
        decoded = _decode_value(f["t"], arr[i] if i < len(arr) else None)
        if decoded is not None:
            out[f["name"]] = decoded
    return out


def encode(type_name: str, obj: dict) -> list[Token]:
    """Encode a named message/def into its positional token array."""
    return _encode_fields(_resolve_def(type_name)["fields"], obj)


def decode(type_name: str, arr: list[Token]) -> dict:
    """Decode a positional token array back into an object for a named type."""
    return _decode_fields(_resolve_def(type_name)["fields"], arr)


def type_tag(type_name: str) -> int:
    tag = MESSAGES.get(type_name, {}).get("tag")
    if tag is None:
        raise ValueError(f'codec: "{type_name}" has no tag (not a top-level message)')
    return tag


def encode_envelope(env: dict) -> list[Token]:
    return _encode_fields(ENVELOPE_FIELDS, env)


def decode_envelope(arr: list[Token]) -> dict:
    return _decode_fields(ENVELOPE_FIELDS, arr)
